lua/feline/statusline_ffi.lua

```diff
@@ -0,0 +1,63 @@
+-- This module provides an interface to Neovim's statusline generator using Lua FFI
+local M = {}
+
+local api = vim.api
+local ffi = require('ffi')
+
+-- Definitions required to use Neovim's build_stl_str_hl function to expand statusline expressions
+-- And also other definitions used by this module
+ffi.cdef [[
+typedef unsigned char char8_t;
+typedef struct window_S win_T;
+extern win_T *curwin;
+typedef struct {} stl_hlrec_t;
+typedef struct {} StlClickDefinition;
+typedef struct {} StlClickRecord;
+int build_stl_str_hl(
+    win_T *wp,
+    char8_t *out,
+    size_t outlen,
+    char8_t *fmt,
+    int use_sandbox,
+    char8_t fillchar,
+    int maxwidth,
+    stl_hlrec_t **hltab,
+    StlClickRecord **tabtab
+);
+
+size_t strlen(const char* str);
+]]
+
+-- Used CType values stored in a local variable to avoid redefining them and improve performance
+local char8_buf_t = ffi.typeof('char8_t[?]')
+local char8_str_t = ffi.typeof('char8_t*')
+
+-- Expand statusline expression, returns a Lua string containing plaintext with only the characters
+-- that'll be displayed in the statusline
+function M.expand_statusline_expr(expr, winid)
+    local stlbuf = char8_buf_t(256)
```

Comment:
You can create the buffer in module level and keep reusing that buffer for better performence .